frontend/catalyst/api_extensions/callbacks.py

```diff
@@ -102,20 +103,39 @@ def hybrid_fn(x):
         kwargs.pop("func")
         return functools.partial(accelerate, **kwargs)
 
-    jitted_fn = jax.jit(func)
+    # If this is a partial, we need to make the tracers part of the input
+    is_partial = isinstance(func, Partial)
+    context = []
+    if is_partial:
+        context = tree_leaves(func)
+        func = tree_map(shaped_abstractify, func)
+
+    def total(context, *args, **kwargs):
+        nonlocal func
+        if is_partial:
+            _tracers, shape = tree_flatten(func)
+            func = tree_unflatten(shape, context)
+            return func(*args, **kwargs)
```

Comment:
The non-local `func` was a bit confusing, maybe this is slightly clearer?
```suggestion

    def total(context, *args, **kwargs):
        if is_partial:
            shape = tree_structure(func)
            func = tree_unflatten(shape, context)
            return func(*args, **kwargs)
```

Reply:
This is wrong. Here you do need `nonlocal` because you are assigning back to `func`. If you instead had

```suggestion
    def total(context, *args, **kwargs):
        if is_partial:
            shape = tree_structure(func)
            func2 = tree_unflatten(shape, context)
            return func2(*args, **kwargs)
```

Then it would work. (I confirmed it by running it locally)

Reply:
These two programs seem equivalent to me, `func` is also a local variable in my version?

Reply:
No. `nonlocal` is still required because we are assigning to `func`. Run the example and you'll get this error: local variable 'func' referenced before assignment (inside the exception block).

Reply:
Ah, because for the other branch func is not defined, that makes sense I guess.

Reply:
Do you want to avoid `nonlocal`? Or is `nonlocal` ok? (Again, from a conceptual point of view, we are calling the same function (modulo bound variables) so I think both of them having the same name is preferrable.

Reply:
`nonlocal` is fine then, maybe just remove the obsolete line 👍 
```suggestion

    def total(context, *args, **kwargs):
        nonlocal func
        if is_partial:
            shape = tree_structure(func)
            func = tree_unflatten(shape, context)
            return func(*args, **kwargs)
```